Include/internal/pycore_weakref.h

```diff
@@ -8,49 +8,76 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+#include "pycore_object.h"           // _Py_REF_IS_MERGED()
 #include "pycore_critical_section.h" // Py_BEGIN_CRITICAL_SECTION()
 
-static inline PyObject* _PyWeakref_GET_REF(PyObject *ref_obj) {
-    assert(PyWeakref_Check(ref_obj));
-    PyWeakReference *ref = _Py_CAST(PyWeakReference*, ref_obj);
-    PyObject *obj = ref->wr_object;
-
-    if (obj == Py_None) {
-        // clear_weakref() was called
-        return NULL;
+static inline int _is_dead(PyObject *obj)
+{
+    if (obj == NULL) {
+        return 0;
     }
-
     // Explanation for the Py_REFCNT() check: when a weakref's target is part
     // of a long chain of deallocations which triggers the trashcan mechanism,
     // clearing the weakrefs can be delayed long after the target's refcount
     // has dropped to zero.  In the meantime, code accessing the weakref will
     // be able to "see" the target object even though it is supposed to be
     // unreachable.  See issue gh-60806.
-    Py_ssize_t refcnt = Py_REFCNT(obj);
-    if (refcnt == 0) {
-        return NULL;
+#if defined(Py_GIL_DISABLED)
+    Py_ssize_t refcount = Py_REFCNT(obj);
+    if (refcount == 0) {
+        // In freethreading CPython, Py_REFCNT is only estimated count,
+        // so need to check shared refcount value too.
+        Py_ssize_t shared = _Py_atomic_load_ssize_relaxed(&obj->ob_ref_shared);
+        return _Py_REF_IS_MERGED(shared);
     }
+    return 0;
```

Comment:
This is a bit tricky.  It's not enough to check that `shared` is merged -- the shared reference count must also be zero. The `Py_REFCNT(obj) == 0` check may seem to imply this, but it doesn't both because it happens earlier in the code (and the reference count may change between the calls) and because the `Py_REFCNT(obj)` check itself is prone to race conditions since it needs to load the local and shared reference counts separately. That's why it's only an estimate.

It should be enough to only check the that the shared reference count is zero and that it is merged.